src/Messages/OWSMessageSender.m

```diff
@@ -697,26 +677,12 @@ - (void)groupSend:(NSArray<SignalRecipient *> *)recipients
     [completionFuture catchDo:^(id failure) {
         // failure from toc_thenAll yeilds an array of failed Futures, rather than the future's failure.
         if ([failure isKindOfClass:[NSArray class]]) {
-            NSArray *errors = (NSArray *)failure;
-            for (TOCFuture *failedFuture in errors) {
-                if (!failedFuture.hasFailed) {
-                    // If at least one send succeeded, don't show message as failed.
-                    // Else user will tap-to-resend to all recipients, including those that already received the
-                    // message.
-                    return successHandler();
-                }
-            }
-
-            // At this point, all recipients must have failed.
-            // But we have all this verbose type checking because TOCFuture doesn't expose type information.
-            id lastError = errors.lastObject;
-            if ([lastError isKindOfClass:[TOCFuture class]]) {
-                TOCFuture *failedFuture = (TOCFuture *)lastError;
-                if (failedFuture.hasFailed) {
-                    id failureResult = failedFuture.forceGetFailure;
+            NSArray *groupSendFutures = (NSArray *)failure;
+            for (TOCFuture *groupSendFuture in groupSendFutures) {
```

Comment:
> The downside is that we'll confuse the situation where multiple recipients do get it but, e.g. the one unregistered group member, does not get it.

You raise an excellent point.  I thought about this overnight and came up with some further changes: https://github.com/WhisperSystems/SignalServiceKit/pull/173/commits/a7df1c18d0eb515bfe34ecb3fc1b05f313811573

* I've moved the "is retryable" flag from a separate boolean in the failure handler to being a property of the error.  This lets us propagate the flag from the inner loop to the outer loop.
* I've added a "error should be ignored for group sends" flag.  So far I'm only using it for the case you mentioned (sending group messages shouldn't appear to fail if one group member unregisters), but I wonder if we should use it for any other kind of error as well.
* I've also made a number of small bug fixes.  I'll call them out with inline comments.